File: InterfaceBuilding/SwiftInterfaceBulding/SwiftInterfaceBulding/VisibilityChangeTester.swift
//
//  VisibilityChangeTester.swift
//  SwiftInterfaceBulding
//
//  Created by Krzysztof Drab on 29/12/2020.
//  Copyright © 2020 Krzysztof Drab. All rights reserved.
//

import Foundation

class VisibilityChangeTester: Tester {
    private var visibleRowsIds: [Int] = []
    private var invisibleRowsIds: [Int] = []
    private var nbOfItems: Int = 0
    private var nbOfChanges: Int = 0
    private var rows: [RowData] = []
    
    override func generateRows() -> [RowData] {
        self.tests = [5, 25, 50, 75, 100, 125, 150, 175, 200];
        self.nbOfItems = 500;
        self.nbOfChanges = tests[testId];
        let totalNbOfItems: Int = self.nbOfChanges + self.nbOfItems;
        self.rows = self.generateDefaultRows(nbOfRows: totalNbOfItems);
        self.visibleRowsIds = Array(0..<(totalNbOfItems));
        self.visibleRowsIds.shuffle()
        self.invisibleRowsIds = Array(self.visibleRowsIds[0..<self.nbOfChanges]);
        self.invisibleRowsIds.forEach { self.rows[$0].visibility = false };
        self.visibleRowsIds = Array(self.visibleRowsIds[self.nbOfChanges...]);
        return self.rows.filter(\.visibility);
    }
    
    override func updateRows() -> [RowData] {
        let startingPoint = Date()
        self.invisibleRowsIds.forEach { self.rows[$0].visibility = true };
        self.visibleRowsIds = self.invisibleRowsIds + self.visibleRowsIds;
        self.invisibleRowsIds = Array(self.visibleRowsIds[(self.visibleRowsIds.count - self.nbOfChanges)...]);
        self.visibleRowsIds = Array(self.visibleRowsIds[0..<(self.visibleRowsIds.count - self.nbOfChanges)]);
        self.invisibleRowsIds.forEach { self.rows[$0].visibility = false };
        self.timeToModify += startingPoint.timeIntervalSinceNow * -1;
        return self.rows.filter(\.visibility);
    }
    
    override func testEnded(fps: Double) {
        self.results.append(TestResult(n: nbOfChanges, timeToModify: timeToModify, fps: fps))
    }
}

File: InterfaceBuilding/flutter_interface_building/lib/visibility_change_tester.dart
import 'dart:io' show Platform;

import 'row_data.dart';
import 'tester.dart';

class VisibilityChangeTester extends Tester {
  List<int> tests = Platform.isIOS
    ? [5, 25, 50, 75, 100, 125, 150, 175, 200]
    : [1, 10, 20, 30, 40, 50, 60, 70, 80];
  int nbOfItems = Platform.isIOS ? 500 : 150;
  int nbOfChanges = 0;
  List<RowData> rows;
  List<int> visibleRowsIds;
  List<int> invisibleRowsIds;

  List<RowData> generateRows() {
    this.nbOfChanges = this.tests[this.testId];
    int totalNbOfItems = this.nbOfChanges + this.nbOfItems;
    this.rows = this.generateDefaultRows(totalNbOfItems);
    this.visibleRowsIds = new List<int>.generate(totalNbOfItems, (i) => i);
    this.visibleRowsIds.shuffle();
    this.invisibleRowsIds = this.visibleRowsIds.sublist(0, this.nbOfChanges);
    this.invisibleRowsIds.forEach((id) => this.rows[id].visibility = false);
    this.visibleRowsIds.removeRange(0, this.nbOfChanges);
    return this.rows.where((e) => e.visibility).toList();
  }

  List<RowData> updateRows() {
    final stopwatch = Stopwatch()..start();
    this.invisibleRowsIds.forEach((id) => this.rows[id].visibility = true);
    this.visibleRowsIds = this.invisibleRowsIds + this.visibleRowsIds;
    this.invisibleRowsIds = this.visibleRowsIds.sublist(this.visibleRowsIds.length - this.nbOfChanges);
    this.visibleRowsIds.removeRange(
      this.visibleRowsIds.length - this.nbOfChanges,
      this.visibleRowsIds.length
    );
    this.invisibleRowsIds.forEach((id) => this.rows[id].visibility = false);
    this.timeToModify += stopwatch.elapsedMilliseconds;
    return this.rows.where((e) => e.visibility).toList();
  }

  void testEnded(double fps) {
    this.results.add(TestResult(this.nbOfChanges, fps, this.timeToModify));
  }
}
